lighthouse-core/gather/gatherers/seo/font-size.js

```diff
@@ -316,16 +254,93 @@ class FontSize extends Gatherer {
+    // give the entire Node object, so DOM.getFlattenedDocument is used. The only connection between a snapshot
+    // Node and an actual Protocol Node is backendId, so that is used to join the two data structures.
+    const snapshot = await passContext.driver.sendCommand('DOMSnapshot.captureSnapshot', {
+      computedStyles: ['font-size'],
+    });
+
+    // Makes the strings access code easier to read.
+    /** @param {number} index */
+    const lookup = (index) => snapshot.strings[index];
+
+    // Locate the document under analysis.
+    // TODO: this needs to use frameId
+    const doc = snapshot.documents.find(doc => lookup(doc.documentURL) === passContext.url);
+
+    // doc is a flattened property list describing all the Nodes in a document, with all string values
+    // deduped in a strings array.
+    // Implementation:
+    // https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/inspector/inspector_dom_snapshot_agent.cc?sq=package:chromium&g=0&l=534
+
+    if (!doc || !doc.nodes.nodeType || !doc.nodes.nodeName || !doc.nodes.backendNodeId
+      || !doc.nodes.nodeValue || !doc.nodes.parentIndex) {
+      throw new Error('Unexpected response from DOMSnapshot.captureSnapshot.');
+    }
+
+    // Not all nodes have computed styles (ex: TextNodes), so doc.layout.* is smaller than doc.nodes.*
+    // doc.layout.nodeIndex maps the index into doc.nodes.* to an index into doc.layout.styles.
+    // nodeIndexToStyleIndex inverses that mapping.
+    /** @type {Map<number, number>} */
+    const nodeIndexToStyleIndex = new Map();
+    for (let i = 0; i < doc.layout.nodeIndex.length; i++) {
+      nodeIndexToStyleIndex.set(doc.layout.nodeIndex[i], i);
+    }
+
+    /** @type {Map<number, {fontSize: number, textLength: number}>} */
+    const backendIdsToPartialFontData = new Map();
+    for (let i = 0; i < doc.nodes.nodeType.length; i++) {
+      const nodeType = doc.nodes.nodeType[i];
+      const nodeValue = lookup(doc.nodes.nodeValue[i]);
+      if (!isNonEmptyTextNode({
+        nodeType,
+        nodeValue,
+        parentNodeName: lookup(doc.nodes.nodeName[doc.nodes.parentIndex[i]]),
+      })) continue;
+
+      const styleIndex = nodeIndexToStyleIndex.get(doc.nodes.parentIndex[i]);
+      if (!styleIndex) continue;
+      const parentStyles = doc.layout.styles[styleIndex];
+      const [fontSizeStringId] = parentStyles;
+      const fontSize = parseInt(lookup(fontSizeStringId), 10);
+      backendIdsToPartialFontData.set(doc.nodes.backendNodeId[i], {
+        fontSize,
+        // TODO: trimming this for a second time. maybe don't?
+        textLength: getNodeTextLength(nodeValue),
+      });
+    }
+
+    const nodes = await getAllNodesFromBody(passContext.driver);
+
+    // backendIdsToPartialFontData will include all Nodes,
+    // but nodes will only contain the Body node and its descendants.
+
+    /** @type {NodeFontData[]} */
+    const failingNodes = [];
```

Comment:
do you think we could move all this into some functions for the different steps you're adding so `afterPass` will be pretty short like before? I think it might help human-parsers figure out what's going on :)

it looks like there's roughly the following stages

1. get the flattened document + snapshot and make sure its all valid
2. create our mappings to be able to tie the two together
3. go through all the nodes and identify the failing ones

is that roughly accurate?

Reply:
Yeah that sounds good. I mostly wanted to wait for PR feedback before chopping up afterPass, just incase the whole thing was moot or something :P